File: 0-Math/67-Add-Binary/67-Add-Binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        stringstream ss;
        int i = a.length() - 1;
        int j = b.length() - 1;
        int carry = 0;
        while (i >= 0 || j >= 0)
        {
            int sum = carry;
            if (i >= 0) 
                sum += (int) a[i] - '0';
            if (j >= 0)
                sum += (int) b[j] - '0';
            ss << sum % 2;
            carry = sum / 2;
            i--;
            j--;
        }
        if (carry)
            ss << carry;
        string s = ss.str();
        reverse(s.begin(), s.end());
        return s;
    }
};

/**
 * Notes: Use an int to store the addition result
 * by using if statement to see whether we need
 * to add another zero. Then we determine the value
 * for the current bit and the carry on digit. 
 * Finally we reverse our string and return. 
 * */

File: 0-Math/67-Add-Binary/67-Add-Binary.java
/**
 * Problem 67: Add Binary
 * Prompt: Given two binary strings, return their sum (also a binary string).
 * The input strings are both non-empty and contains only characters 1 or 0.
 * Date: 07/19/2020
 */
class Solution {
    public String addBinary(String a, String b) {
        if (b.length() > a.length())
        {
            String t = a;
            a = b;
            b = t;
        }
        char[] arr = a.toCharArray();
        boolean carry = false;
        int j = a.length() - 1;
        for (int i = b.length() - 1; i >= 0; --i)
        {
            if (b.charAt(i) == '1')
            {
                if (arr[j] == '1')
                {
                    if (!carry)
                    {
                        carry = true;
                        arr[j] = '0';
                    }
                }
                else
                {
                    if (!carry)
                        arr[j] = '1';
                }
            }
            else if (carry)
            {
                if (arr[j] == '1')
                    arr[j] = '0';
                else
                {
                    arr[j] = '1';
                    carry = false;
                }
            }
            j--;
        }
        while (carry && j >= 0)
        {
            if (arr[j] == '1')
                arr[j] = '0';
            else
            {
                carry = false;
                arr[j] = '1';
            }
            j--;
        }
        if (!carry)
            return String.valueOf(arr);
        return '1' + String.valueOf(arr);
            
    }
}

/**
 * Notes: Do addition step by step from end to front. Need
 * to using different branches for different conditions on the 
 * addition of the one bit. 
 * In the end, if there are leftover carried number we need to 
 * concatenate that '1'. 
 */